server/internal/xorg/xorg.c

```diff
@@ -96,21 +159,66 @@ void XButton(unsigned int button, int down) {
   }
 }
 
+KeyCode XkbKeysymToKeycode(KeySym keysym) {
+  XkbDescPtr xkb;
+  XkbStateRec state;
+  unsigned keycode;
+  Display *dpy = getXDisplay();
+
+  xkb = XkbGetMap(dpy, XkbAllComponentsMask, XkbUseCoreKbd);
+  if (!xkb)
+    return 0;
+
+  XkbGetState(dpy, XkbUseCoreKbd, &state);
+
+  for (keycode = xkb->min_key_code;
+       keycode <= xkb->max_key_code;
+       keycode++) {
+    KeySym cursym;
+    unsigned int mods;
+    XkbTranslateKeyCode(xkb, keycode, state.compat_state, &mods, &cursym);
+    if (cursym == keysym)
+      break;
+  }
+
+  if (keycode > xkb->max_key_code)
+    keycode = 0;
+
+  XkbFreeKeyboard(xkb, XkbAllComponentsMask, True);
+
+  return keycode;
+}
+
 void XKey(unsigned long key, int down) {
   if (key != 0) {
     Display *display = getXDisplay();
-    KeyCode code = XKeysymToKeycode(display, key);
-
-    // Map non-existing keysyms to new keycodes
-    if(code == 0) {
-      int min, max, numcodes;
-      XDisplayKeycodes(display, &min, &max);
-      XGetKeyboardMapping(display, min, max-min, &numcodes);
-
-      code = (max-min+1)*numcodes;
-      KeySym keysym_list[numcodes];
-      for(int i=0;i<numcodes;i++) keysym_list[i] = key;
-      XChangeKeyboardMapping(display, code, numcodes, keysym_list, 1);
+//    KeyCode code = XKeysymToKeycode(display, key);
+
+    int code;
+
+    if (down) {
+      code = searchItem(key);
+      if (code == 0) {
+        // XKeysymToKeycode() doesn't respect state, so we have to use
+        // something slightly more complex
+        code = XkbKeysymToKeycode(key);
+        // Map non-existing keysyms to new keycodes
+        if(code == 0) {
+          int min, max, numcodes;
+          XDisplayKeycodes(display, &min, &max);
+          XGetKeyboardMapping(display, min, max-min, &numcodes);
+
+          code = (max-min+1)*numcodes;
+          KeySym keysym_list[numcodes];
+          for(int i=0;i<numcodes;i++) keysym_list[i] = key;
+            XChangeKeyboardMapping(display, code, numcodes, keysym_list, 1);
+        }
+
+	insertAtLast(key, code);
+      }
+    } else {
+      code = searchItem(key);
```

Comment:
Edge case, when key was never pressed but only released is not handled. That might not happen, but since we are controlling it from within a browser, you can theoretically do keypress outside of neko, come with cursor to neko and do only release there. When `XTestFakeKeyEvent` receives `0`, it crashes.